tests/gateways/disk/test_lock.py

```diff
@@ -0,0 +1,110 @@
+# Copyright (C) 2012 Anaconda, Inc
+# SPDX-License-Identifier: BSD-3-Clause
+
+import builtins
+import sys
+
+import pytest
+from pytest import CaptureFixture, MonkeyPatch
+from pytest_mock import MockerFixture
+
+from conda.common.compat import on_win
+from conda.exceptions import LockError
+from conda.gateways.disk.lock import (
+    _lock_impl,
+    lock,
+)
+
+
+# A function mocking the signature of the real import method in the builtein module
+def monkey_import_ImportError(name, globals=None, locals=None, fromlist=(), level=0):
+    if name in ("msvcrt", "fcntl"):
+        raise ImportError("Mocked import error")
+
+
+def test_locking_not_supported(monkeypatch: MonkeyPatch, tmp_path):
+    tmp_file = tmp_path / "testfile"
+    tmp_file.write_bytes(b"test")
+    monkeypatch.delitem(sys.modules, "fcntl", raising=False)
+    monkeypatch.delitem(sys.modules, "msvcrt", raising=False)
+    monkeypatch.setattr(builtins, "__import__", monkey_import_ImportError)
+
+    with pytest.raises(ImportError):
+        lock(tmp_file)
+
+
+@pytest.mark.skipif(not on_win, reason="windows-specific test")
+def test_LockError_raised_windows(
+    mocker: MockerFixture, monkeypatch: MonkeyPatch, tmp_path
+):
+    tmp_file = tmp_path / "testfile"
+    tmp_file.write_bytes(b"test")
+    mocker.patch("msvcrt.locking", side_effect=OSError)
+    monkeypatch.setattr("conda.gateways.disk.lock.LOCK_ATTEMPTS", 1)
+    with pytest.raises(LockError):
+        with tmp_file.open("r+b") as f:
+            with lock(f):
+                pass
+
+
+@pytest.mark.skipif(on_win, reason="non windows test")
+def test_LockError_raised_not_windows(
+    mocker: MockerFixture, monkeypatch: MonkeyPatch, tmp_path
+):
+    tmp_file = tmp_path / "testfile"
+    tmp_file.write_bytes(b"test")
+
+    mocker.patch("fcntl.lockf", side_effect=OSError)
+    monkeypatch.setattr("conda.gateways.disk.lock.LOCK_ATTEMPTS", 1)
+    with pytest.raises(LockError):
+        with tmp_file.open("r+b") as f:
+            with lock(f):
+                pass
```

Comment:
These can be combined:

```suggestion
def test_LockError_raised(
    mocker: MockerFixture,
    monkeypatch: MonkeyPatch,
    tmp_path: Path,
):
    (tmp_file := tmp_path / "testfile").touch()

    mocker.patch("msvcrt.locking" if on_win else "fcntl.lockf", side_effect=OSError)
    monkeypatch.setattr("conda.gateways.disk.lock.LOCK_ATTEMPTS", 1)

    with pytest.raises(LockError):
        with tmp_file.open("r+b") as fh, lock(fh):
            pass
```